drivers/usb/usbehci_new/hardware.h

```diff
@@ -0,0 +1,417 @@
+    ULONG Reserved1                 : 1;
+    ULONG IsoSchedulingThreshold    : 4;
+    ULONG ExtCapabilitiesPointer    : 8; // (EECP)
+    ULONG Reserved2                 : 16;
+  };
+  ULONG AsULONG;
+} EHCI_HC_CAPABILITY_PARAMS;
+
+C_ASSERT(sizeof(EHCI_HC_CAPABILITY_PARAMS) == sizeof(ULONG));
+
+typedef struct _EHCI_HC_CAPABILITY_REGISTERS {
+  UCHAR RegistersLength; // RO
+  UCHAR Reserved; // RO
+  USHORT InterfaceVersion; // RO
+  EHCI_HC_STRUCTURAL_PARAMS StructParameters; // RO
+  EHCI_HC_CAPABILITY_PARAMS CapParameters; // RO
+  UCHAR CompanionPortRouteDesc[8]; // RO
+} EHCI_HC_CAPABILITY_REGISTERS, *PEHCI_HC_CAPABILITY_REGISTERS;
+
+typedef union _EHCI_USB_COMMAND {
+  struct {
+    ULONG Run                        : 1;
+    ULONG Reset                      : 1;
+    ULONG FrameListSize              : 2;
+    ULONG PeriodicEnable             : 1;
+    ULONG AsynchronousEnable         : 1;
+    ULONG InterruptAdvanceDoorbell   : 1;
+    ULONG LightResetHC               : 1; // optional
+    ULONG AsynchronousParkModeCount  : 2; // optional
+    ULONG Reserved1                  : 1;
+    ULONG AsynchronousParkModeEnable : 1; // optional
+    ULONG Reserved2                  : 4;
+    ULONG InterruptThreshold         : 8;
+    ULONG Reserved3                  : 8;
+  };
+  ULONG AsULONG;
+} EHCI_USB_COMMAND;
+
+C_ASSERT(sizeof(EHCI_USB_COMMAND) == sizeof(ULONG));
+
+typedef union _EHCI_USB_STATUS {
+  struct {
+    ULONG Interrupt               : 1;
+    ULONG ErrorInterrupt          : 1;
+    ULONG PortChangeDetect        : 1;
+    ULONG FrameListRollover       : 1;
+    ULONG HostSystemError         : 1;
+    ULONG InterruptOnAsyncAdvance : 1;
+    ULONG Reserved1               : 6;
+    ULONG HCHalted                : 1;
+    ULONG Reclamation             : 1;
+    ULONG PeriodicStatus          : 1;
+    ULONG AsynchronousStatus      : 1;
+    ULONG Reserved2               : 16;
+  };
+  ULONG AsULONG;
+} EHCI_USB_STATUS;
+
+C_ASSERT(sizeof(EHCI_USB_STATUS) == sizeof(ULONG));
+
+#define EHCI_INTERRUPT_MASK  0x3F 
+
+typedef union _EHCI_INTERRUPT_ENABLE {
+  struct {
+    ULONG Interrupt               : 1;
+    ULONG ErrorInterrupt          : 1;
+    ULONG PortChangeInterrupt     : 1;
+    ULONG FrameListRollover       : 1;
+    ULONG HostSystemError         : 1;
+    ULONG InterruptOnAsyncAdvance : 1;
+    ULONG Reserved                : 26;
+  };
+  ULONG AsULONG;
+} EHCI_INTERRUPT_ENABLE;
+
+C_ASSERT(sizeof(EHCI_INTERRUPT_ENABLE) == sizeof(ULONG));
+
+#define EHCI_LINE_STATUS_K_STATE_LOW_SPEED 1
+#define EHCI_PORT_OWNER_COMPANION_CONTROLLER 1
+
+typedef union _EHCI_PORT_STATUS_CONTROL {
+  struct {
+    ULONG CurrentConnectStatus    : 1;
+    ULONG ConnectStatusChange     : 1;
+    ULONG PortEnabledDisabled     : 1;
+    ULONG PortEnableDisableChange : 1;
+    ULONG OverCurrentActive       : 1;
+    ULONG OverCurrentChange       : 1;
+    ULONG ForcePortResume         : 1;
+    ULONG Suspend                 : 1;
```

Comment:
Is this _Suspend Status_ or _Suspend Using LX_?

Reply:
There's only one "Suspend" in the port status and control register, and it's called "Suspend". See section 2.3.9 in the EHCI spec.

Reply:
> [Suspend using L1 – R/W. Default = 0b. 0b = Suspend using L2; 1b = Suspend using L1 (LPM). When this bit is set to a one and a non-zero value is specified in the Device Address field the host controller will generate an LPM Token to enter the L1 state whenever software writes a one to the Suspend bit, as well as L1 exit timing during any device- or host-initiated resume. When set to zero the host controller will employ the legacy (L2) suspend mechanism. Software should only set this bit when the device attached immediately downstream of this root port supports L1 transitions.](https://www.intel.com/content/dam/www/public/us/en/documents/technical-specifications/ehci-v1-1-addendum.pdf)

Reply:
Okay. This is neither of those two. Bit 9 is Reserved1, bits 23/24 are part of Reserved2. Bit7 is still just "Suspend" even in the document you linked.